### DISCUSSION TITLE
Is `update_focused_file` running in a synchronized/blocking way or async/non-blocking way?

### DISCUSSION THREAD

**Discussion by linrongbin16** [nvim-tree/nvim-tree.lua#2275] [Q&A] [answered]
Hi, in 2 months ago, I'm using 'nvim-tree.lua' in the company's git repository which I'm working for.

The git repository is a super large monorepo, which contains maybe hundreds of directories, hundreds thousands of files.

When switching buffers, nvim-tree is stucking on finding the focused file. When I disable this option, it will be fine.

In the same time, an other plugin [neo-tree.nvim](https://github.com/nvim-neo-tree/neo-tree.nvim) is using async/non-blocking way, so it don't stuck, and maybe 0.5~1 seconds later, it will focused on the file I'm editing.

After 2 months, is there any updates?

**Comment by alex-courtis** [comment]
Almost all git operations are async/background #1974

Please raise a [bug report](https://github.com/nvim-tree/nvim-tree.lua/issues/new?assignees=&labels=bug&projects=&template=bug_report.yml) with
1. clean room configuration
2. instructions on how to replicate
3. a very large repo that we may replicate with. If you can't share your own, https://github.com/qmk/qmk_firmware is one that will stress nvim-tree

**Reply by linrongbin16** [reply]
thanks for reply, but what I want to say is:

when nvim-tree is tracing to the focused file, is the tracing (e.g. finding path) sync or async?

**Reply by alex-courtis** [reply, marked as answer]
That is synchronous, however is a cheap operation.

Please add some [profiling](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#diagnostic-logging) and report your results, so that we may better investigate this problem:
```lua
  log = {
    enable = true,
    truncate = true,
    types = {
      profile = true,
      git = true,
    },
  },
```

**Reply by linrongbin16** [reply]
Oh, I'm sorry, I'm afraid since the log contains some sensitive data from the company I'm working for.
So I cannot directly upload the log here. Maybe just let's ignore this thing. Thank you anyway.

**Reply by alex-courtis** [reply]
Many users have scrubbed their logs before uploading. `man sed`

All I'm interested in are the find file and draw timings e.g.

```
[2023-06-19 16:05:06] [profile] START find file /home/alex/src/nvim-tree/master/README.md
[2023-06-19 16:05:06] [profile] START draw
[2023-06-19 16:05:06] [profile] END   draw 1ms
[2023-06-19 16:05:06] [profile] END   find file /home/alex/src/nvim-tree/master/README.md 1ms
```

Take a look and see if you see any large times.

**Reply by linrongbin16** [reply]
I removed the sensitive data, and also replaced some sensitive words with other common words:

[nvim-tree.log](https://github.com/nvim-tree/nvim-tree.lua/files/11784948/nvim-tree.log)

I found two timings are big:
```
[2023-06-19 14:20:40] [profile] END   refresh_nodes_for_path /Users/rlin/company/node_modules/appname/components/entity 1906ms

[2023-06-19 14:20:40] [profile] END   find file /Users/rlin/company/node_modules/appname/components/entity/EntityViewSection.jsx 1928ms
```

I'm working on macbooking pro, hardware performance should not be an issue:
<img width="280" alt="image" src="https://github.com/nvim-tree/nvim-tree.lua/assets/6496887/4f7f3cd9-9697-4172-b681-c66ace5d3c20">

**Reply by alex-courtis** [reply]
This one is interesting:
`END   find file /Users/rlin/company/eng/componentname/app/graphql/componentname/mutations/entity_doing_mutations/doing_entity_total_to_ca.rb 2179ms`

Is that git ignored? Reminds me of #1931 which is still outstanding but has workarounds.

Please try putting your slow directories (hopefully git ignored) in `:help nvim-tree.filesystem_watchers.ignore_dirs`.

**Reply by linrongbin16** [reply]
No, that's not git ignored files, it's just in one of the source code directories in the super big git repo.

**Reply by alex-courtis** [reply]
What was the result of using ignore_dirs?

Let's raise a proper bug report for this:
* clean room configuration
* sanitised contents of your ~/.gitconfig

Please attempt to replicate against another large repo e.g. https://github.com/qmk/qmk_firmware
It will be very difficult to diagnose if we cannot replicate the problem.

**Reply by linrongbin16** [reply]
OK, I will try use an open source large repo. will do later.

**Reply by linrongbin16** [reply]
Hi @alex-courtis ,

finally I use linux kernel git repo: https://github.com/torvalds/linux.git

It's large enough to show some logs (but actually compare to the git monorepo in the company I'm working, still not big enough), here's the log:

[nvim-tree.log](https://github.com/nvim-tree/nvim-tree.lua/files/11794997/nvim-tree.log)

The `refresh_nodes_for_path` and `find files` (in my git monorepo, they're about 2 secs):
```
line-672:
[2023-06-20 09:36:13] [profile] END   refresh_nodes_for_path /Users/rlin/github/torvalds/linux/arch/powerpc/boot/dts/fsl 641ms

line-675
[2023-06-20 09:36:13] [profile] END   find file /Users/rlin/github/torvalds/linux/arch/powerpc/boot/dts/fsl/qoriq-dma-1.dtsi 700ms
```

**Reply by alex-courtis** [reply]
Thanks @linrongbin16 I'll get to this one on the weekend.

**Reply by alex-courtis** [reply]
Raised #2293, let's follow this there.

**Reply by linrongbin16** [reply]
really thanks for your work
